interfaces/builtin/docker.go

```diff
@@ -0,0 +1,176 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var dockerPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the Docker daemon. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+# Allow sockets
+/{,var/}run/docker.sock rw,
+/{,var/}run/docker/     rw,
+/{,var/}run/docker/**   mrwklix,
+/{,var/}run/runc/       rw,
+/{,var/}run/runc/**     mrwklix,
+
+# Wide read access to /proc, but somewhat limited writes for now
+@{PROC}/ r,
+@{PROC}/** r,
+@{PROC}/[0-9]*/attr/exec w,
+@{PROC}/sys/net/** w,
+@{PROC}/[0-9]*/cmdline r,
+@{PROC}/[0-9]*/oom_score_adj w,
+
+# Wide read access to /sys
+/sys/** r,
+# Limit cgroup writes a bit
+/sys/fs/cgroup/*/docker/   rw,
+/sys/fs/cgroup/*/docker/** rw,
+
+# Allow tracing ourself (especially the "runc" process we create)
+ptrace (trace) peer=@{profile_name},
+
+# Docker needs a lot of caps, but limits them in the app container
+capability,
+
+# Docker does all kinds of mounts all over the filesystem
+/dev/mapper/control rw,
+/dev/mapper/docker* rw,
+/dev/loop* r,
+/dev/loop[0-9]* w,
+mount,
+umount,
+pivot_root,
+/.pivot_root*/ rw,
```

Comment:
Ok, I did some testing, and this is still necessary, but I did make it slightly more specific (`/.pivot_root[0-9]*/`) and added a better comment to explain what's happening:
```
# After doing a pivot_root using <graph-dir>/<container-fs>/.pivot_rootNNNNNN,
# Docker removes the leftover /.pivot_rootNNNNNN directory (which is now
# relative to "/" instead of "<graph-dir>/<container-fs>" thanks to pivot_root)
pivot_root,
/.pivot_root[0-9]*/ rw,
```